### PR TITLE
TASK: Use argument unpacking instead of call_user_func_array

### PR DESCRIPTION
**Pull request by albe**
This is supported since PHP 5.6 and is actually faster than call_user_func_array at least since PHP 7.
Just be sure to not use it in places where the method is static or a Closure.
This also deprecates `ObjectAccess::instantiateClass()` (retrospectively). It will be removed with Flow 7.

### CODE REVIEW

File: Neos.Flow/Classes/Aop/AdvicesTrait.php [hunk 1]
```diff
@@ -52,7 +52,8 @@ public function Flow_Aop_Proxy_invokeJoinPoint(\Neos\Flow\Aop\JoinPointInterface
         if (__CLASS__ !== $joinPoint->getClassName()) {
             return parent::Flow_Aop_Proxy_invokeJoinPoint($joinPoint);
         }
-        if (isset($this->Flow_Aop_Proxy_methodIsInAdviceMode[$joinPoint->getMethodName()])) {
+        $methodName = $joinPoint->getMethodName();
+        if (isset($this->Flow_Aop_Proxy_methodIsInAdviceMode[$methodName])) {
```
Comment (albe) [inline Neos.Flow/Classes/Aop/AdvicesTrait.php:56]:
```suggestion
        if (isset($this->Flow_Aop_Proxy_methodIsInAdviceMode[$joinPoint->getMethodName()])) {
```
Let's maybe not do this here?

File: Neos.Flow/Classes/Mvc/Controller/ActionController.php [hunk 2]
```diff
@@ -462,7 +462,7 @@ protected function callActionMethod()
         $validationResult = $this->arguments->getValidationResults();
 
         if (!$validationResult->hasErrors()) {
-            $actionResult = call_user_func_array([$this, $this->actionMethodName], $preparedArguments);
+            $actionResult = $this->{$this->actionMethodName}(...$preparedArguments);
```
Comment (albe) [inline Neos.Flow/Classes/Mvc/Controller/ActionController.php:465]: These cases are debatable... the `->{$this->*MethodName}()` syntax is not super nice. I still slightly prefer it over the `call_user_func_array`, but I'm not gonna fight for it.

**Comment by kitsunet** [comment]
Good call, somehow I thought I opened a PR for that before, but that might still sit around here somewehere, so good to have it!

File: Neos.Utility.ObjectHandling/Classes/ObjectAccess.php [hunk 3]
```diff
@@ -435,45 +435,13 @@ public static function buildSetterMethodName(string $propertyName): string
...
-     * For less than 7 arguments `new` is used, for more a `ReflectionClass` is created
-     * and `newInstanceArgs` is used.
-     *
-     * Note: this should be used sparingly, just calling `new` yourself or using Dependency
-     * Injection are most probably better alternatives.
-     *
      * @param string $className
      * @param array $arguments
      * @return object
+     * @deprecated directly use "new $className(...$arguments)" instead
```
Comment (albe) [inline Neos.Utility.ObjectHandling/Classes/ObjectAccess.php:441]: We could postpone the deprecation to 5.2 if wanted. Does it hurt though?

File: Neos.Flow/Classes/Aop/AdvicesTrait.php (hunk 1 above)
Comment (kitsunet) [reply Neos.Flow/Classes/Aop/AdvicesTrait.php:56]: Right, how did that sneak in

Comment (albe) [reply Neos.Flow/Classes/Aop/AdvicesTrait.php:56]: I was about to replace the following line with `return self::$methodName(...$joinPoint->getMethodArguments());` but then was unsure and it turned out in a half-change.

File: Neos.Flow/Classes/Mvc/Controller/ActionController.php (hunk 2 above)
Comment (kitsunet) [reply Neos.Flow/Classes/Mvc/Controller/ActionController.php:465]: Mmm, IDK, we could do something like: 

```
$actionMethodName = $this->actionMethodName;
$this->$actionMethodName(...$preparedArguments);
```

But I don't think that's much easier readable... I think it's fine, it's odd anywys and might need a revisit at some point.

File: Neos.Utility.ObjectHandling/Classes/ObjectAccess.php (hunk 3 above)
Comment (kitsunet) [reply Neos.Utility.ObjectHandling/Classes/ObjectAccess.php:441]: nah, let's deprecatae this right away. We can always decide to leave it in for another round of major...

**Comment by albe** [comment]
```
1) Neos\Flow\Tests\Unit\SignalSlot\DispatcherTest::dispatchPassesTheSignalArgumentsToTheSlotMethod
Error: Cannot unpack array with string keys
```
Ah-ha! Edge-cases be edgy. Well... wrapping all arguments everywhere in an `array_values()` before unpacking is counter-effective so I guess... we have to revert back to *some* call_user_func_array?

File: Neos.Flow/Classes/Aop/AdvicesTrait.php (hunk 1 above)
Comment (kitsunet) [reply Neos.Flow/Classes/Aop/AdvicesTrait.php:56]: sure!

**Comment by kitsunet** [comment]
We could just do array_values in that particular place?

File: Neos.Flow/Classes/Aop/AdvicesTrait.php (hunk 1 above)
Comment (albe) [reply Neos.Flow/Classes/Aop/AdvicesTrait.php:56]: Note: because the `$methodName` might be a static method (?)

**Comment by albe** [comment]
> We could just do array_values in that particular place?

Was about to do that, then noticed that `$object` might be the class name in case of static slot method. Reverted to `call_user_func_array` to avoid an additional `if (is_string($object)) {` condition

**Comment by kitsunet** [comment]
right, fair enough....

File: Neos.Flow/Classes/Aop/AdvicesTrait.php [hunk 4]
```diff
@@ -52,7 +52,8 @@ public function Flow_Aop_Proxy_invokeJoinPoint(\Neos\Flow\Aop\JoinPointInterface
...
             return call_user_func_array(['self', $joinPoint->getMethodName()], $joinPoint->getMethodArguments());
```
Comment (albe) [inline Neos.Flow/Classes/Aop/AdvicesTrait.php:57]:
```suggestion
            return self::$methodName(...$joinPoint->getMethodArguments());
```
Alternative to above revert suggestion. You decide @kitsunet ;P

**Comment by albe** [comment]
IMO looks good to merge

File: Neos.Flow/Classes/Session/Aspect/LazyLoadingAspect.php [hunk 5]
```diff
@@ -118,7 +118,7 @@ public function callMethodOnOriginalSessionObject(JoinPointInterface $joinPoint)
         if ($this->sessionOriginalInstances[$objectName] === $proxy) {
             return $joinPoint->getAdviceChain()->proceed($joinPoint);
         } else {
-            return call_user_func_array([$this->sessionOriginalInstances[$objectName], $methodName], $joinPoint->getMethodArguments());
+            return $this->sessionOriginalInstances[$objectName]->$methodName(...$joinPoint->getMethodArguments());
```
Comment (albe) [inline Neos.Flow/Classes/Session/Aspect/LazyLoadingAspect.php:121]:
```suggestion
            $arguments = array_values($joinPoint->getMethodArguments());
            return $this->sessionOriginalInstances[$objectName]->$methodName(...$arguments);
```

File: Neos.Flow/Classes/Aop/AdvicesTrait.php [hunk 6]
```diff
@@ -52,8 +52,9 @@ public function Flow_Aop_Proxy_invokeJoinPoint(\Neos\Flow\Aop\JoinPointInterface
         if (__CLASS__ !== $joinPoint->getClassName()) {
             return parent::Flow_Aop_Proxy_invokeJoinPoint($joinPoint);
         }
-        if (isset($this->Flow_Aop_Proxy_methodIsInAdviceMode[$joinPoint->getMethodName()])) {
-            return call_user_func_array(['self', $joinPoint->getMethodName()], $joinPoint->getMethodArguments());
+        $methodName = $joinPoint->getMethodName();
+        if (isset($this->Flow_Aop_Proxy_methodIsInAdviceMode[$methodName])) {
+            return self::$methodName(...$joinPoint->getMethodArguments());
```
Comment (albe) [inline Neos.Flow/Classes/Aop/AdvicesTrait.php:57]:
```suggestion
            $arguments = array_values($joinPoint->getMethodArguments());
            return self::$methodName(...$arguments);
```

**Comment by albe** [comment]
@kitsunet wdyt? good?

**Comment by albe** [comment]
Let's do this... what can possibly go wrong? 😄 If to does go wrong, I'll be ready to hit revert.
